docs/source/corev_hw_loop.rst

```diff
@@ -43,76 +43,84 @@ Hardware Loop constraints
 
 The HWLoop constraints are:
 
--  Start and End address of an HWLoop must be word aligned
+-  Start and End address of an HWLoop must be word aligned.
+
+-  End address of an HWLoop must point to the instruction just after the last one of the loop body.
 
 -  HWLoop body must contain at least 3 instructions.
-   An illegal exception is raised otherwise.
+   An illegal instruction exception is raised otherwise.
 
 -  No Compressed instructions (RVC) allowed in the HWLoop body.
-   An illegal exception is raised otherwise.
+   An illegal instruction exception is raised otherwise.
 
--  No uncoditional jump instructions allowed in the HWLoop body.
-   An illegal exception is raised otherwise.
+-  No unconditional jump instructions allowed in the HWLoop body.
+   An illegal instruction exception is raised otherwise.
 
--  No coditional branch instructions allowed in the HWLoop body.
-   An illegal exception is raised otherwise.
+-  No conditional branch instructions allowed in the HWLoop body.
+   An illegal instruction exception is raised otherwise.
 
 -  No privileged instructions (mret, dret, ecall, wfi) allowed in the HWLoop body, except for ebreak.
-   An illegal exception is raised otherwise.
+   An illegal instruction exception is raised otherwise.
 
 -  No memory ordering instructions (fence, fence.i) allowed in the HWLoop body.
-   An illegal exception is raised otherwise.
+   An illegal instruction exception is raised otherwise.
 
 -  The End address of the outermost HWLoop (#1) must be at least 2
    instructions further than the End address innermost HWLoop (#0),
    i.e. HWLoop[1].endaddress >= HWLoop[0].endaddress + 8
-   An illegal exception is raised otherwise.
+   An illegal instruction exception is raised otherwise.
```

Comment:
as above

Reply:
Like the "at least 3 instructions", those 2 constraints are harder to tackle in hardware. It would require to add resources (adders/substractors) just to check these wrong cases. Not really nice to keep CV32E40P small and power efficient.

In my previous life, these cases were only managed by constraining compiler and were not checked by HW ...

Reply:
None required if no exception.
For SW, they should not generate those cases so no pb.
For Reference Model, emulate an instruction exception.
For verification, those cases should not be verified and so not generated. If it would happen anyway, there will be a difference for sure as Reference Model will generate an exception and the HW will not ...

Reply:
mmm I wouldn't go this direction, but I do not have strong opinions either.

We won't definitively sell it as "safe" then, but even if you do nothing in HW, the HW won't work, so we should write AT LEAST that the behaviour is undefined, and we should remove the statements that we raise an exception, as we won't - did I understand this correctly?

Reply:
ok then you can say that the reference model does it, but I would not, I mean, in theory, the reference model should behave as the DUT, but I let @MikeOpenHWGroup chime in here

Reply:
Actually, I always say that the DUT should behave as the reference model.  :wink:

Seriously, this is an important topic.  In the general case there are three conditions:
1. **Illegal events that result in either unpredictable behavior and/or an unrecoverable state.**
The testbench will be constrained to avoid generating these events.  The reference model will not check for these events.

2. **Illegal events that cause an exception.**
The testbench will be able to generate these events and will have functional coverage to ensure that it does.  The reference model will also generate an exception.

3. **Illegal events that are automatically recoverable in the DUT.**
The testbench will be able to generate these events and will have functional coverage to ensure that it does.  The reference model will recover in the same way as the DUT.

Reply:
I generally agree with the first statement **except** when the reference model is not only used just for verification but also as an execution model for software guys when developing their application.
This is the case when this RM is embedded in a Virtual Platform.
If the RM is not generating an exception for case 1. then you are finding the issue only when simulating on RTL or FPGA.
If the compiler is not generating this case no pb but what if "someone" writes assembly code not respecting hw loops constraints?
I know I know writing assembly code !!!